WTFReportBacktrace fails to print backtrace and crashes on ASAN

kkinnunen-apple opened the pull request:
#### 5b5a2dc989fa65afa3eef1c084b690a01fa17813
<pre>
WTFReportBacktrace fails to print backtrace and crashes on ASAN
<a href="https://bugs.webkit.org/show_bug.cgi?id=245826">https://bugs.webkit.org/show_bug.cgi?id=245826</a>
rdar://problem/100557350

Reviewed by Mark Lam and Yusuke Suzuki.

WTFReportBacktrace() would use StackTrace in the &quot;borrowed&quot; mode.
The borrowed mode stack pointer would be obtained by reading
uninitialized !m_capacity. This would cause read to random memory.

Fix by removing the StackTrace borrowed mode. When printing
the backtrace, use the new class StackTracePrinter which
is able to use the void* array WTFGetStacktrace returns.

Move the size/isBorrowed members out of the union that was
intended to optimize skipping of the two initial, skipped
frames. This was relying on two counts of UB:
- reading of non-active member of the union:
  Filled the m_stack by writing into one union m_skippedFrame0
  making it active, but then reading from passive union
  member m_size.

- Writing into one member, and expecting the overflowing
  of the write being defined to match the subsequent member:
  Oversized write into union member m_skippedFrame0
  was assumend to spill over to member of other, subsequent union
  member m_stack. This is not well-defined.

The above UB invocations were done to trying to save two
void* worth of memory in the StackTrace, while avoiding
copying of the void* array.

Instead, first read the stack trace into the memory storage
of StackTrace, then overwrite the initial portion with the
real class data.

Fix bugs in the unborrowed mode, where
captureStackTrace(maxFrames, framesToSkip)
would not actually skip the framesToSkip amount of frames.
If left unfixed, we could not assert that the borrowed
mode fix did not accidentally cause regressions to the
unborrowed mode.

Move the printing related StackTrace::m_prefix to
StackTracePrinter::m_prefix. This way StackTrace size shrinks.

Test the equivalent of WTFReportBacktrace in a scenario
that caused the uninitialized read.

Test the captureStackTrace().

(WTF::StackTrace::instanceSize):
(WTF::StackTrace::captureStackTrace):
* Source/WTF/wtf/StackTrace.h:
(WTF::StackTrace::StackTrace):
(WTF::StackTrace::stack const):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/StackTraceTest.cpp: Added.
(TestWebKitAPI::TEST):
(TestWebKitAPI::expectEqualFrames):

Canonical link: <a href="https://commits.webkit.org/255164@main">https://commits.webkit.org/255164@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ed7653d9919eeeb40270ec68d80c29f6fc97fa4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91460 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/556 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22112 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101186 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161200 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/575 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83808 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97549 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97118 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/389 "Found 1 new test failure: fast/forms/ios/file-upload-panel-capture.html (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78181 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27349 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82313 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/81973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70384 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35594 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16001 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33380 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17094 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3586 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37180 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39888 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39100 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36211 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/WTF/wtf/StackTrace.h:
```diff
@@ -129,33 +130,20 @@ class StackTrace {
-    // top 2 frame which is of no interest. Setting up the fields layout this way
-    // allows us to capture the stack in place and minimize space wastage.
+    const bool m_isBorrowed { false };
+    int m_size;
```
Let's move `m_initialFrame` up here like so:
```
    const bool m_isBorrowed { false };
    int m_initialFrame : 24;
    int m_size { 0 };
```
We know that `m_initialFrame` will never be that large in practice anyway.  So, we can just squeeze it in here and save us 8 bytes (due to alignment).

Let's also make sure to default `m_size` to 0.

I'd prefer the `int`s to be `unsigned` instead, but I think I went with `int`s in the original code because some other API is using `int`, and I went with the easy route to avoid casting.  It can probably remain `int` and still work just fine.

MenloDorian commented on Source/WTF/wtf/StackTrace.h:
```diff
@@ -129,33 +130,20 @@ class StackTrace {
-            void* m_skippedFrame1;
+            int m_initialFrame;
+            void* m_stack[1];
         };
```
This reduces to just `void* m_stack[1];` (see above).

MenloDorian commented on Source/WTF/wtf/StackTrace.h:
```diff
@@ -50,21 +50,22 @@ class PrintStream;
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    WTF_EXPORT_PRIVATE static std::unique_ptr<StackTrace> captureStackTrace(int maxFrames, int framesToSkip = 0);
+    WTF_EXPORT_PRIVATE NEVER_INLINE static std::unique_ptr<StackTrace> captureStackTrace(int maxFrames, int framesToSkip = 0);
```
You also need to make `WTFGetBacktrace` NEVER_INLINE.  The "+ 2" that we add to `framesToSkip` is because of `captureStackTrace` and `WTFGetBacktrace`.

MenloDorian commented on Source/WTF/wtf/StackTrace.cpp:
```diff
@@ -44,31 +45,27 @@ void WTFGetBacktrace(void** stack, int* size)
-        RELEASE_ASSERT(numberOfFrames >= framesToSkip);
+    std::unique_ptr<StackTrace> trace(new (NotNull, fastMalloc(instanceSize(numberOfFrames))) StackTrace());
+    WTFGetBacktrace(trace->m_stack, &numberOfFrames);
+    if (numberOfFrames > framesToSkip) {
```
Let's move this `WTFGetBacktrace` call inside the `if (numberOfFrames > framesToSkip)` check.  We don't need to call it if we don't intend to use any of the captured frames anyway.

MenloDorian reviewed (approved):
r=me with suggestions for improvements.

Constellation reviewed (approved):
r=me

MenloDorian commented on Source/WTF/wtf/Assertions.h:
```diff
@@ -230,7 +230,7 @@ WTF_EXPORT_PRIVATE bool WTFWillLogWithLevel(WTFLogChannel*, WTFLogLevel);
 
 WTF_EXPORT_PRIVATE void WTFGetBacktrace(void** stack, int* size);
```
`WTFGetBacktrace` needs to be NEVER_INLINE too.  We're counting on this as one of the frames that we'll skip.

MenloDorian commented on Source/WTF/wtf/StackTrace.h:
```diff
@@ -50,27 +51,42 @@ class PrintStream;
+    {
+    }
+
+    size_t m_size;
+    size_t m_initialFrame;
```
You should double check (with a `static_assert`) that `sizeof(size_t) == sizeof(void*)` on `CPU(ARM64_32)`.  Otherwise, use `uintptr_t` here?

kkinnunen-apple replied:
Checked in the .cpp implementation that uses the fact.

MenloDorian reviewed (approved):
r=me with fixes.

kkinnunen-apple committed:
WTFReportBacktrace fails to print backtrace and crashes on ASAN
https://bugs.webkit.org/show_bug.cgi?id=245826
rdar://problem/100557350

Reviewed by Mark Lam and Yusuke Suzuki.

WTFReportBacktrace() would use StackTrace in the "borrowed" mode.
The borrowed mode stack pointer would be obtained by reading
uninitialized !m_capacity. This would cause read to random memory.

Fix by removing the StackTrace borrowed mode. When printing
the backtrace, use the new class StackTracePrinter which
is able to use the void* array WTFGetStacktrace returns.

Move the size/isBorrowed members out of the union that was
intended to optimize skipping of the two initial, skipped
frames. This was relying on two counts of UB:
- reading of non-active member of the union:
  Filled the m_stack by writing into one union m_skippedFrame0
  making it active, but then reading from passive union
  member m_size.

- Writing into one member, and expecting the overflowing
  of the write being defined to match the subsequent member:
  Oversized write into union member m_skippedFrame0
  was assumend to spill over to member of other, subsequent union
  member m_stack. This is not well-defined.

The above UB invocations were done to trying to save two
void* worth of memory in the StackTrace, while avoiding
copying of the void* array.

Instead, first read the stack trace into the memory storage
of StackTrace, then overwrite the initial portion with the
real class data.

Fix bugs in the unborrowed mode, where
captureStackTrace(maxFrames, framesToSkip)
would not actually skip the framesToSkip amount of frames.
If left unfixed, we could not assert that the borrowed
mode fix did not accidentally cause regressions to the
unborrowed mode.

Move the printing related StackTrace::m_prefix to
StackTracePrinter::m_prefix. This way StackTrace size shrinks.

Test the equivalent of WTFReportBacktrace in a scenario
that caused the uninitialized read.

Test the captureStackTrace().

(WTF::StackTrace::instanceSize):
(WTF::StackTrace::captureStackTrace):
* Source/WTF/wtf/StackTrace.h:
(WTF::StackTrace::StackTrace):
(WTF::StackTrace::stack const):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/StackTraceTest.cpp: Added.
(TestWebKitAPI::TEST):
(TestWebKitAPI::expectEqualFrames):

Canonical link: https://commits.webkit.org/255164@main
